test/e2e/singlecluster/statefulset_test.go

```diff
@@ -133,5 +129,96 @@ var _ = ginkgo.Describe("Stateful set integration", func() {
 			util.ExpectObjectToBeDeleted(ctx, k8sClient, statefulSet, true)
 			util.ExpectObjectToBeDeletedWithTimeout(ctx, k8sClient, createdWorkload, false, util.LongTimeout)
 		})
+
+		ginkgo.It("should delete all pods on scale down to zero", func() {
+			statefulSet := statefulsettesting.MakeStatefulSet("sts", ns.Name).
+				Image(util.E2eTestSleepImage, []string{"10m"}).
+				Request(corev1.ResourceCPU, "100m").
+				Replicas(3).
+				Queue(lq.Name).
+				Obj()
+			wlLookupKey := types.NamespacedName{Name: statefulset.GetWorkloadName(statefulSet.Name), Namespace: ns.Name}
+
+			ginkgo.By("Create StatefulSet", func() {
+				gomega.Expect(k8sClient.Create(ctx, statefulSet)).To(gomega.Succeed())
+			})
+
+			ginkgo.By("Waiting for replicas is ready", func() {
+				gomega.Eventually(func(g gomega.Gomega) {
+					createdStatefulSet := &appsv1.StatefulSet{}
+					g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(statefulSet), createdStatefulSet)).To(gomega.Succeed())
+					g.Expect(createdStatefulSet.Status.ReadyReplicas).To(gomega.Equal(int32(3)))
+				}, util.LongTimeout, util.Interval).Should(gomega.Succeed())
+			})
+
+			createdWorkload := &kueue.Workload{}
+			ginkgo.By("Check workload is created", func() {
+				gomega.Expect(k8sClient.Get(ctx, wlLookupKey, createdWorkload)).To(gomega.Succeed())
+			})
+
+			ginkgo.By("Scale down replicas to zero", func() {
+				gomega.Eventually(func(g gomega.Gomega) {
+					createdStatefulSet := &appsv1.StatefulSet{}
+					g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(statefulSet), createdStatefulSet)).To(gomega.Succeed())
+					createdStatefulSet.Spec.Replicas = ptr.To[int32](0)
+					g.Expect(k8sClient.Update(ctx, createdStatefulSet)).To(gomega.Succeed())
+				}, util.LongTimeout, util.Interval).Should(gomega.Succeed())
+			})
+
+			ginkgo.By("Try to scale up immediately", func() {
```

Comment:
This might be racy if the STS gets scaled down quickly. Please add 2 test cases:
- one which does scale up after ReadyReplicas = 0 (the main use case for most users)
- one which does scale up when ReadyReplicas < 3 (and retries)

Reply:
Both cases already tested on this test.

Reply:
> one which does scale up after ReadyReplicas = 0 (the main use case for most users)

This case tested on:
https://github.com/kubernetes-sigs/kueue/blob/eb4214030a96e9e629d36e634263df64c59273c1/test/e2e/singlecluster/statefulset_test.go#L168-L173

> one which does scale up when ReadyReplicas < 3 (and retries)

This case tested on:
https://github.com/kubernetes-sigs/kueue/blob/eb4214030a96e9e629d36e634263df64c59273c1/test/e2e/singlecluster/statefulset_test.go#L201-L208

Reply:
Done